File: app/src/main/java/com/foreverrafs/radiocore/util/Serializers.java
package com.foreverrafs.radiocore.util;

import com.foreverrafs.radiocore.model.News;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import org.joda.time.DateTime;

import java.lang.reflect.Type;

import timber.log.Timber;

public class Serializers {

    public static class NewsDeserializer implements JsonDeserializer<News> {
        @Override
        public News deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            String headline = ((JsonObject) json).get("headline").getAsString();
            String content = ((JsonObject) json).get("content").getAsString();
            String category = ((JsonObject) json).get("category").getAsString();
            String imageUrl = ((JsonObject) json).get("imageUrl").getAsString();
            String dateStr = ((JsonObject) json).get("date").getAsString();
            DateTime date = DateTime.now();

            try {
                date = DateTime.parse(dateStr);
            } catch (Exception exception) {
                Timber.i(exception);
            }

            return new News(headline, content, date, category, imageUrl);
        }
    }

    public static class NewsSerializer implements JsonSerializer<News> {

        @Override
        public JsonElement serialize(News src, Type typeOfSrc, JsonSerializationContext context) {
            return null;
        }
    }
}

File: app/src/main/java/com/foreverrafs/radiocore/util/Serializers.kt
package com.foreverrafs.radiocore.util

import com.foreverrafs.radiocore.model.News
import com.google.gson.*
import org.joda.time.DateTime
import timber.log.Timber
import java.lang.reflect.Type

class Serializers {
    class NewsDeserializer : JsonDeserializer<News> {
        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): News {
            val headline = (json as JsonObject)["headline"].asString
            val content = json["content"].asString
            val category = json["category"].asString
            val imageUrl = json["imageUrl"].asString
            val dateStr = json["date"].asString
            var date = DateTime.now()
            try {
                date = DateTime.parse(dateStr)
            } catch (exception: Exception) {
                Timber.i(exception)
            }
            return News(headline, content, date!!, category, imageUrl)
        }
    }

    class NewsSerializer : JsonSerializer<News?> {
        override fun serialize(src: News?, typeOfSrc: Type, context: JsonSerializationContext): JsonElement? {
            return null
        }
    }
}